mirgecom/profiling.py

```diff
@@ -138,48 +92,113 @@ def __init__(self, queue, allocator=None) -> None:
         self.profile_events = []
         self.profile_results = {}
         self.kernel_stats = {}
+        self.logmgr = logmgr
 
-        init_pyopencl_array_monkey_patch()
+        cl.array.ARRAY_KERNEL_EXEC_HOOK = self.array_kernel_exec_hook
 
     def __del__(self):
         """Release resources and undo monkey patching."""
         del self.profile_events[:]
         self.profile_results.clear()
         self.kernel_stats.clear()
 
-        del_pyopencl_array_monkey_patch()
+        cl.array.ARRAY_KERNEL_EXEC_HOOK = None
+
+    def array_kernel_exec_hook(self, knl, queue, gs, ls, *actual_args, wait_for):
+        """Extract data from the elementwise array kernel."""
+        evt = knl(queue, gs, ls, *actual_args, wait_for=wait_for)
+
+        name = knl.function_name
+        knl.name = knl.function_name  # "name" is used by the LoopKernels
+
+        args_tuple = tuple(
+            (arg.size)
+            for arg in actual_args if isinstance(arg, cl.array.Array))
+
+        try:
+            ignore = self.kernel_stats[knl][args_tuple]  # noqa
+        except KeyError:
+            nbytes = 0
+            nops = 0
+            for arg in actual_args:
+                if isinstance(arg, cl.array.Array):
+                    nbytes += arg.size * arg.dtype.itemsize
+                    nops += arg.size
+            res = SingleCallKernelProfile(time=0, flops=nops, bytes_accessed=nbytes,
+                                footprint_bytes=nbytes)
+            self.kernel_stats.setdefault(knl, {})[args_tuple] = res
+
+        if self.logmgr and f"{name}_time" not in self.logmgr.quantity_data:
+            self.logmgr.add_quantity(KernelProfile(self, name))
+
+        self.profile_events.append(ProfileEvent(evt, knl, args_tuple))
+
+        return evt
 
     def _finish_profile_events(self) -> None:
```

Comment:
IMO this method is a bit mis-named: It transfers data from `profile_events` to `profile_results`. Maybe a different name could be found that reflects that better?

Reply:
renamed to `_wait_and_transfer_profile_events`